From 3f628ae0a482b977e88fd668d9ae0c8eb4d72a2a Mon Sep 17 00:00:00 2001
From: Manikanta Mylavarapu <quic_mmanikan@quicinc.com>
Date: Mon, 20 Nov 2023 17:34:26 +0530
Subject: [PATCH] firmware: scm: add support for q6 break at start

For debugging purpose, we may need to break q6
at start. This patch adds support to break q6.

Change-Id: Ib7aac07a308dede1d767c777d57aca05f4dacc48
Signed-off-by: Manikanta Mylavarapu <quic_mmanikan@quicinc.com>
---
 drivers/firmware/qcom_scm.c | 20 ++++++++++++++++++++
 include/linux/qcom_scm.h    |  3 +++
 2 files changed, 23 insertions(+)

diff --git a/drivers/firmware/qcom_scm.c b/drivers/firmware/qcom_scm.c
index 57557c35a70b..d912970f063d 100644
--- a/drivers/firmware/qcom_scm.c
+++ b/drivers/firmware/qcom_scm.c
@@ -720,6 +720,26 @@ int qcom_scm_pas_mem_setup(u32 peripheral, phys_addr_t addr, phys_addr_t size)
 }
 EXPORT_SYMBOL(qcom_scm_pas_mem_setup);
 
+int qcom_scm_break_q6_start(u32 reset_cmd_id)
+{
+	int ret;
+	struct qcom_scm_desc desc = {
+		.svc = QCOM_SCM_SVC_PIL,
+		.cmd = reset_cmd_id,
+		.arginfo = QCOM_SCM_ARGS(1),
+		.args[0] = QCOM_BREAK_Q6,
+		.owner = ARM_SMCCC_OWNER_SIP,
+	};
+	struct qcom_scm_res res;
+
+	ret = qcom_scm_call(__scm->dev, &desc, &res);
+	if (ret || res.result[0])
+		return ret ? : res.result[0];
+
+	return 0;
+}
+EXPORT_SYMBOL(qcom_scm_break_q6_start);
+
 /**
  * qcom_scm_pas_auth_and_reset() - Authenticate the given peripheral firmware
  *				   and reset the remote processor
diff --git a/include/linux/qcom_scm.h b/include/linux/qcom_scm.h
index c608dcf361c0..fa0e2b876cdc 100644
--- a/include/linux/qcom_scm.h
+++ b/include/linux/qcom_scm.h
@@ -21,6 +21,8 @@
 #define QTI_SCM_SVC_FUSE		0x8
 #define QTI_TRYBIT			BIT(12)
 
+#define QCOM_BREAK_Q6			0x1
+
 #define MAX_FUSE_ADDR_SIZE		0x8
 struct fuse_payload {
 	uint32_t fuse_addr;
@@ -231,6 +233,7 @@ void qcom_scm_pas_metadata_release(struct qcom_scm_pas_metadata *ctx);
 extern int qcom_scm_pas_mem_setup(u32 peripheral, phys_addr_t addr,
 				  phys_addr_t size);
 extern long qcom_scm_is_feature_available(u32 feature_id);
+extern int qcom_scm_break_q6_start(u32 reset_cmd_id);
 extern int qcom_scm_pas_auth_and_reset(u32 peripheral);
 extern int qcom_scm_pas_shutdown(u32 peripheral);
 extern bool qcom_scm_pas_supported(u32 peripheral);
-- 
2.34.1

